### ISSUE TITLE
[Question][Hw4] Possible optimizations.

### ISSUE THREAD

**Issue by Lee-Janggun** [prosyslab-classroom/cs492-program-reasoning#89]
Hi, I'm currently working on topdown synthesis, and have implemented a normalization algorithm with memonization (simply using a set) to get rid of common programs. 

For the order of programs to evaluate, I am evaluating small grounded programs first, followed by programs with the least non-terminal nodes.

For me, even with normalization, the improvement in performance is minimal. In particular, for `example4.sl`, while the set of programs to evaluate does seem to increase a lot slower, (about 100,000 cases in the memonized set after 10 seconds), there is still too much to reasonably do in under the minute time limit. 

I am wondering if there is a normalization case that I am missing, or if we are expected to implement a more sophisticated pruning method. 

If there are any hints the TAs or someone else can tell me, I'll be very thankful. I am having a lot of fun doing this assignment, but I feel that I may have hit a roadblock.

FYI, the normalization strategies I have done are as follows.

* Const evaluation.
* Variable elimination for possible cases (i.e, (+ (- 6 x) x) becomes 6).
* Forcing of left associativity. (i.e, only (+ (+ e1 e2) e3)) is allowed.
* Forcing 0 to be the LHS of comparison operators.
* Other minor ones so the above normalizations works as intended

**Comment by retroinspect** [comment]
Maybe TDP is crucial I guess.

**Comment by KihongHeo** [comment]
Hi. 
It seems that Janggun's ranking suddenly went up. So maybe you found a way.

BTW, I encourage every student to develop their own optimizations based on the lectures. So it is not appropriate to share all the secret recipe. Instead, a few students who get top scores will be invited to present their strategies in class.

A tip: sometimes low-level implementation issues may cause performance bottleneck. A common mistake is non-trail recursive function call. Try to find such things in you code. 

Hope everybody enjoys.

**Comment by Lee-Janggun** [comment]
I optimized my bottomup code, which helped pass some more cases. 

I'll look over my implementation to see if there are any more bottlenecks, thanks for the tip!

**Comment by KihongHeo** [comment]
Perfect!

Also I am very happy with that all the participants so far defeated me.

**Comment by kohs100** [comment]
I'm sorry to comment on an issue that's already been completed, but doesn't forcing 0 to be the LHS of comparison operators result in increasing the size of the resulting program?

**Comment by Lee-Janggun** [comment]
Yes, but it also introduces more normalization opportunities, and I had other ways to deal with such changes that increase size.
